File: Helper/Core/Actions/Actions.m
//
// --------------------------------------------------------------------------
// Actions.m
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2020
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#import "Actions.h"
#import "TouchSimulator.h"
#import "SharedUtility.h"
#import "ModificationUtility.h"
#import "HelperUtility.h"
#import "Modifiers.h"
#import "Remap.h"
#import "Constants.h"

#import "Logging.h"
/*
@import Carbon;
#import "CGSHotKeys.h"
*/
#import "SymbolicHotKeys.h"
#import <Carbon/Carbon.h>

@implementation Actions

+ (void)executeActionArray:(NSArray *)actionArray phase:(MFActionPhase)phase {
    
    DDLogDebug(@"Executing action array: %@, phase: %@", actionArray, @(phase));
    
    if (phase == kMFActionPhaseEnd) {
        return; /// TODO: Actually implement actions with different phases
    }
               
    for (NSDictionary *actionDict in actionArray) {
        
        MFStringConstant actionType = actionDict[kMFActionDictKeyType];
    
        if ([actionType isEqualToString:kMFActionDictTypeSymbolicHotkey]) {
            
            MFSymbolicHotkey shk = ((NSNumber *)actionDict[kMFActionDictKeyGenericVariant]).intValue;
            [SymbolicHotKeys post:(CGSSymbolicHotKey)shk];
            
        } else if ([actionType isEqualToString:kMFActionDictTypeNavigationSwipe]) {
            
            /// Universal Back and Forward
            ///     See `Universal Back and Forward.md`
            ///     Implementation Notes:
            ///         - This is similar to the hack in Scroll.m where we zoom more for `org.chromium.Chromium` and other chromium browsers.
            ///         - HACK: [Aug 2025] We consider this a 'hack' since we're getting the bundleID of the app under the mouse pointer in a slow, hacky way without any caching, since we don't have a 'proper' way to doing app-specific stuff, yet. (Began working on that in the`app-specific` branch IIRC)
            ///     TODOs:
            ///         - TODO: Rename the kMF constants to reflect that this is about going Back and Forward, not about navigationSwipes.
            ///             [Aug 2025] Not doing that for the 3.0.6 hotfix, since it would require increasing the config-version, and I'm too lazy for that.
            ///     Bugs:
            ///         - macOS `Mouse Shortcuts` interfere if we simulate MB 4/5.
            ///             See: https://github.com/noah-nuebling/mac-mouse-fix/issues/1521
            ///             Solution Ideas: - Send at different eventTap - User feedback ala Swish - Simulate navigationSwipes where possible - Some other hacks to keep macOS from diverting the events.
            ///
            
            /// Dispatch to mainThread
            ///     [Aug 2025] Run on the mainThread since `MFEmulateNSMenuItemRemapping()` uses TIS API stuff which wants to run on the mainThread. (And this method is called by Buttons.swift which seems to call on the mainThread for some triggers (observed: .release), but calls on `com.nuebling.mac-mouse-fix.buttons` queue for other triggers (observed: .levelExpired))
            ///         Idea: Perhaps we could optimize by only dispatching to mainThread if we really end up calling TIS APIs.
            dispatch_async(dispatch_get_main_queue(), ^{
            
                #define fail() ({ assert(false); goto endof_universalBackForward; })       /** [Aug 2025] We'll catch the failures during development, so simple is fine */
                {
                    NSString *dirString = actionDict[kMFActionDictKeyGenericVariant];
                    BOOL isleft  = [dirString isEqualToString: kMFNavigationSwipeVariantLeft];
                    BOOL isright = [dirString isEqualToString: kMFNavigationSwipeVariantRight];
                
                    if (phase == kMFActionPhaseEnd)     fail(); /// [Aug 2025] We'll have to update this when we support separate handling of button-up and button-down events
                    if (!(isleft || isright))           fail(); /// [Aug 2025] `kMFNavigationSwipeVariantUp` and `kMFNavigationSwipeVariantDown` are no longer supported (and they were never used AFAIK. `kMFActionDictTypeNavigationSwipe` serves as 'Universal Back and Forward' now. (We should rename it.)
                    
                    /// Choose the `bfmethod`
                    ///     Mnemonic: (method) for going (b)ack and (f)orward
                    
                    NSString *bundleID = [HelperUtility appUnderMousePointerWithEvent: NULL].bundleIdentifier; /// [Aug 2025] Should we query frontmost app or app-under-mouse-pointer? I think navigation swipes only work when the app is frontmost *and* the mouse pointer is over the desired view. Meanwhile the keyboard shortcuts dont depend on mouse pointer position.
                    #define isbundle(bundleid)  [bundleID hasPrefix: @bundleid]                             /** [Aug 2025] Using `hasPrefix:` to also catch other release channels like "com.google.Chrome.canary", or maybe forks that didn't bother to change the bundleID. (?) */
                    {
                        /// Fallback if we can't retrieve a bundleID
                        ///     Note: [Aug 2025] Not sure when this occurs. Maybe non-app executables or certain cross-platform apps? `bfmethod_mouseButton` seems most useful.
                        if (bundleID == nil || bundleID.length == 0)    goto bfmethod_mouseButton;
                        
                        /// navigationSwipe overrides from linearmouse
                        ///     Note: linearmouse uses navigationSwipes for Firefox, but Firefox supports MB 4/5 now. (See `https://stackoverflow.com/a/68532003`). [Aug 2025]
                        if (isbundle("com.operasoftware.Opera"))        goto bfmethod_navigationSwipe;
                        if (isbundle("com.binarynights.ForkLift"))      goto bfmethod_navigationSwipe;
                        
                        /// From mac-mouse-fix issues
                        if (isbundle("org.zotero.zotero"))              goto bfmethod_commandBracket;   /// Behaves like `bfmethod_commandBracket` in Preview when viewing PDFs. [Aug 2025]
                        if (isbundle("com.apple.systempreferences"))    goto bfmethod_commandBracket;
                        if (isbundle("com.apple.AppStore"))             goto bfmethod_commandBracket;
                        if (isbundle("com.adobe.Acrobat.Pro"))          goto bfmethod_commandLeftRightArrow; /// In Acrobat, `bfmethod_commandLeftRightArrow` behaves like `bfmethod_commandBracket` in Preview. Seems pretty useful. [Aug 2025]
                        
                        /// Other
                        if (isbundle("dev.warp.Warp"))                  goto bfmethod_commandBracket;
                        
                        /// Other Apple
                        ///     Note: None of the modern Catalyst/SwiftUI Apple apps support navigationSwipes. Perhaps `bfmethod_commandBracket` should be the 'default' for them instead of `bfmethod_navigationSwipe`.
                        if (isbundle("com.apple.Music"))                goto bfmethod_commandBracket;
                        if (isbundle("com.apple.iCal"))                 goto bfmethod_commandLeftRightArrow;
                        if (isbundle("com.apple.AddressBook"))          goto bfmethod_commandBracket;
                        if (isbundle("com.apple.Notes"))                goto bfmethod_optionCommandBracket;
                        if (isbundle("com.apple.freeform"))             goto bfmethod_optionCommandBracket;
                        if (isbundle("com.apple.TV"))                   goto bfmethod_commandBracket;
                        if (isbundle("com.apple.iBooksX"))              goto bfmethod_commandBracket;
                        if (isbundle("com.apple.Preview"))              goto bfmethod_commandBracket;
                    
                        /// Default
                        if (isbundle("com.apple."))                     goto bfmethod_navigationSwipe; /// Default to navigation swipes for Apple apps
                        else                                            goto bfmethod_mouseButton;     /// Default to MB 4/5 simulation for non-apple apps
                    }
                    #undef isbundle
                    
                    /// Define the `bfmethods`
                    
                    #define bfmethod(bfmethod_label) \
                        goto endof_bfmethods; bfmethod_label: {}; DDLogDebug(@"Actions.m: NavigationSwipe: Posting " #bfmethod_label);
                    {
                        bfmethod(bfmethod_mouseButton) {
                            if (isleft)     [ModificationUtility postMouseButtonClicks: 4 nOfClicks: 1];
                            else            [ModificationUtility postMouseButtonClicks: 5 nOfClicks: 1];
                        }
                        bfmethod(bfmethod_navigationSwipe) {
                            if (isleft)     [TouchSimulator postNavigationSwipeEventWithDirection: kIOHIDSwipeLeft];
                            else            [TouchSimulator postNavigationSwipeEventWithDirection: kIOHIDSwipeRight];
                        }
                        bfmethod(bfmethod_commandBracket) {
                            MFVKCAndFlags *shortcut = MFEmulateNSMenuItemRemapping((isleft ? kVK_ANSI_LeftBracket : kVK_ANSI_RightBracket), kCGEventFlagMaskCommand);
                            postKeyboardShortcut(shortcut.vkc,  (CGSModifierFlags)shortcut.modifierMask);
                        }
                        bfmethod(bfmethod_commandLeftRightArrow) {
                            if (isleft)     postKeyboardShortcut(kVK_LeftArrow,  (CGSModifierFlags)kCGEventFlagMaskCommand);
                            else            postKeyboardShortcut(kVK_RightArrow, (CGSModifierFlags)kCGEventFlagMaskCommand);
                        }
                        bfmethod(bfmethod_optionCommandBracket) {
                            MFVKCAndFlags *shortcut  = MFEmulateNSMenuItemRemapping((isleft ? kVK_ANSI_LeftBracket : kVK_ANSI_RightBracket), (kCGEventFlagMaskAlternate|kCGEventFlagMaskCommand));
                            postKeyboardShortcut(shortcut.vkc,  (CGSModifierFlags)shortcut.modifierMask);
                        }
                    }
                    endof_bfmethods: {}
                    #undef bfmethod
                }
                endof_universalBackForward: {}
                #undef fail
            });
            
        } else if ([actionType isEqualToString:kMFActionDictTypeSmartZoom]) {
            
            [TouchSimulator postSmartZoomEvent];
            
        } else if ([actionType isEqualToString:kMFActionDictTypeKeyboardShortcut]) {
            
            NSNumber *keycode = actionDict[kMFActionDictKeyKeyboardShortcutVariantKeycode];
            NSNumber *flags = actionDict[kMFActionDictKeyKeyboardShortcutVariantModifierFlags];
            postKeyboardShortcut(keycode.intValue, flags.intValue);
            
        } else if ([actionType isEqualToString:kMFActionDictTypeSystemDefinedEvent]) {
            
            NSNumber *type = actionDict[kMFActionDictKeySystemDefinedEventVariantType];
            NSNumber *flags = actionDict[kMFActionDictKeySystemDefinedEventVariantModifierFlags];
            
            postSystemDefinedEvent(type.unsignedIntValue, flags.unsignedIntValue);
            
        } else if ([actionType isEqualToString:kMFActionDictTypeMouseButtonClicks]) {
            
            NSNumber *button = actionDict[kMFActionDictKeyMouseButtonClicksVariantButtonNumber];
            NSNumber *nOfClicks = actionDict[kMFActionDictKeyMouseButtonClicksVariantNumberOfClicks];
            [ModificationUtility postMouseButtonClicks:button.intValue nOfClicks:nOfClicks.intValue];
        
        } else if ([actionType isEqualToString:kMFActionDictTypeAddModeFeedback]) {
            NSMutableDictionary *payload = ((NSMutableDictionary *)actionDict.mutableCopy);
            [payload removeObjectForKey:kMFActionDictKeyType];
            /// ^ Payload has the kMFRemapsKeyTrigger and kMFRemapsKeyModificationPrecondition keys.
            /// It is almost a valid remaps table entry.
            /// All that the main app has to do with the payload in order to make it a valid entry of the remap table's
            ///  dataModel is to add the kMFRemapsKeyEffect key and corresponding values
            [Remap sendAddModeFeedback:payload];
            
        }
    }
}

#pragma mark - System defined events

static void postSystemDefinedEvent(MFSystemDefinedEventType type, NSEventModifierFlags modifierFlags) {
    /// The timestamps, and location, and even the keyUp event seem to be unnecessary. Just trying stuff to try and fix weird bug where music is louder for a few seconds after starting it with a systemDefinedEvent.
    ///     Edit: The bug where the music is too loud for a few seconds also happens when using the keyboard, so the issue is not on our end.
    
    CGEventTapLocation tapLoc = kCGSessionEventTap;
    
    NSPoint loc = NSEvent.mouseLocation;
    
    NSInteger data = 0;
    data = data | kMFSystemDefinedEventBase;
    data = data | (type << 16);
    
    NSInteger downData = data;
    NSInteger upData = data | kMFSystemDefinedEventPressedMask;
    
    /// Post key down
    
    NSTimeInterval ts = [ModificationUtility nsTimeStamp];
    NSEvent *e = [NSEvent otherEventWithType:14 location:loc modifierFlags:modifierFlags timestamp:ts windowNumber:-1 context:nil subtype:8 data1:downData data2:-1];
    
    CGEventPost(tapLoc, e.CGEvent); /// `.CGEvent` "Returns an autoreleased CGEvent" -> Make sure there's an autoreleasepool when calling this!
    
    /// Post key up
    ts = [ModificationUtility nsTimeStamp];
    e = [NSEvent otherEventWithType:14 location:loc modifierFlags:modifierFlags timestamp:ts windowNumber:-1 context:nil subtype:8 data1:upData data2:-1];
    
    CGEventPost(tapLoc, e.CGEvent);
}

#pragma mark - Keyboard shortcuts

static void postKeyboardShortcut(CGKeyCode keyCode, CGSModifierFlags modifierFlags) {
    
    DDLogDebug(@"postKeyboardShortcut: Posting shortcut with %@", vardesc(keyCode, modifierFlags));
    
    CGEventTapLocation tapLoc = kCGSessionEventTap;

    /// Create key events
    CGEventRef keyUp = CGEventCreateKeyboardEvent(NULL, keyCode, false);
    CGEventRef keyDown = CGEventCreateKeyboardEvent(NULL, keyCode, true);
    CGEventSetFlags(keyDown, (CGEventFlags)modifierFlags);
    CGEventSetFlags(keyUp, (CGEventFlags)modifierFlags);
    
    /// Fix up keyboard type [Aug 2025]
    ///     Explanation: [Aug 2025] When I attach 2 keyboards to my Mac, one ANSI, one JIS, then `CGEventCreateKeyboardEvent()` seems to not match the keyboard type retrieved any other way we know (MFKeyboardTypeCurrent(), LMGetKbdType(), LMGetKbdLast(), CGEventSourceCreate()).
    ///         Even when you pass a CGEventSource with the desired keyboardType, CGEventCreateKeyboardEvent() just overrides it. It seems like a bug in CoreGraphics.
    ///         In practise this messes up the 'Universal Back and Forward' feature we're building, cause `MFEmulateNSMenuItemRemapping()` assumes `MFKeyboardTypeCurrent()` when calculating the vkc that will trigger the shortcut.
    ///         Observed on: macOS 15.5, 2018 Mac Mini, [Aug 2025]
    ///
    ///     ! Keep this when merging with __EventLoggerForBrad__.
    CGEventSetIntegerValueField(keyDown, kCGKeyboardEventKeyboardType, MFKeyboardTypeCurrent()); /// Keep in sync with `MFEmulateNSMenuItemRemapping()` to make 'Universal Back and Forward' feature work [Aug 2025]
    CGEventSetIntegerValueField(keyUp,   kCGKeyboardEventKeyboardType, MFKeyboardTypeCurrent());
    
    /// Create modifier restore event
    ///  (Restoring original modifier state the way `postKeyboardEventsForSymbolicHotkey()` does leads to issues with triggering Spotlight)
    ///      (Sometimes triggered Siri instead)
//    CGEventFlags originalModifierFlags = CGEventGetFlags(CGEventCreate(NULL));
    CGEventRef modEvent = CGEventCreate(NULL);
//    CGEventSetFlags(modEvent, originalModifierFlags);
    
    /// Send key events
    CGEventPost(tapLoc, keyDown);
    CGEventPost(tapLoc, keyUp);
    CGEventPost(tapLoc, modEvent);
    
    CFRelease(keyDown);
    CFRelease(keyUp);
    CFRelease(modEvent);
}

@end

File: Helper/Core/Actions/Actions.swift
//
// --------------------------------------------------------------------------
// Actions.swift
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2020
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#if IS_HELPER

import Cocoa
import Carbon

@objc public enum MFActionPhase: Int {
    case start
    case end
    case combined
}

public let kMFActionPhaseStart = MFActionPhase.start
public let kMFActionPhaseEnd = MFActionPhase.end
public let kMFActionPhaseCombined = MFActionPhase.combined

@objc(Actions)
public class Actions: NSObject {
    
    @objc public static func executeActionArray(_ actionArray: NSArray, phase: MFActionPhase) {
        DDLogDebug("Executing action array: \(actionArray), phase: \(phase.rawValue)")
        
        if phase == kMFActionPhaseEnd {
            return
        }
        
        for action in actionArray {
            guard let actionDict = action as? NSDictionary else { continue }
            guard let actionType = actionDict[kMFActionDictKeyType] as? String else { continue }
            
            if actionType == kMFActionDictTypeSymbolicHotkey {
                if let num = actionDict[kMFActionDictKeyGenericVariant] as? NSNumber {
                    let shk = num.intValue
                    SymbolicHotKeys.post(CGSSymbolicHotKey(UInt32(shk)))
                }
            } else if actionType == kMFActionDictTypeNavigationSwipe {
                
                // 通用前进后退
                DispatchQueue.main.async {
                    guard let dirString = actionDict[kMFActionDictKeyGenericVariant] as? String else { return }
                    let isLeft = (dirString == kMFNavigationSwipeVariantLeft)
                    let isRight = (dirString == kMFNavigationSwipeVariantRight)
                    
                    if phase == kMFActionPhaseEnd { return }
                    if !(isLeft || isRight) { return }
                    
                    let bundleID = HelperUtility.appUnderMousePointer(with: nil)?.bundleIdentifier
                    
                    func isBundle(_ prefix: String) -> Bool {
                        guard let bid = bundleID else { return false }
                        return bid.hasPrefix(prefix)
                    }
                    
                    if bundleID == nil || bundleID!.isEmpty {
                        bfmethod_mouseButton()
                        return
                    }
                    
                    if isBundle("com.operasoftware.Opera") { bfmethod_navigationSwipe(); return }
                    if isBundle("com.binarynights.ForkLift") { bfmethod_navigationSwipe(); return }
                    
                    if isBundle("org.zotero.zotero") { bfmethod_commandBracket(); return }
                    if isBundle("com.apple.systempreferences") { bfmethod_commandBracket(); return }
                    if isBundle("com.apple.AppStore") { bfmethod_commandBracket(); return }
                    if isBundle("com.adobe.Acrobat.Pro") { bfmethod_commandLeftRightArrow(); return }
                    
                    if isBundle("dev.warp.Warp") { bfmethod_commandBracket(); return }
                    
                    if isBundle("com.apple.Music") { bfmethod_commandBracket(); return }
                    if isBundle("com.apple.iCal") { bfmethod_commandLeftRightArrow(); return }
                    if isBundle("com.apple.AddressBook") { bfmethod_commandBracket(); return }
                    if isBundle("com.apple.Notes") { bfmethod_optionCommandBracket(); return }
                    if isBundle("com.apple.freeform") { bfmethod_optionCommandBracket(); return }
                    if isBundle("com.apple.TV") { bfmethod_commandBracket(); return }
                    if isBundle("com.apple.iBooksX") { bfmethod_commandBracket(); return }
                    if isBundle("com.apple.Preview") { bfmethod_commandBracket(); return }
                    
                    if isBundle("com.apple.") {
                        bfmethod_navigationSwipe()
                    } else {
                        bfmethod_mouseButton()
                    }
                    
                    func bfmethod_mouseButton() {
                        DDLogDebug("Actions.swift: NavigationSwipe: Posting bfmethod_mouseButton")
                        if isLeft {
                            ModificationUtility.postMouseButtonClicks(MFMouseButtonNumber(4), nOfClicks: 1)
                        } else {
                            ModificationUtility.postMouseButtonClicks(MFMouseButtonNumber(5), nOfClicks: 1)
                        }
                    }
                    
                    func bfmethod_navigationSwipe() {
                        DDLogDebug("Actions.swift: NavigationSwipe: Posting bfmethod_navigationSwipe")
                        if isLeft {
                            TouchSimulator.postNavigationSwipeEvent(withDirection: IOHIDSwipeMask(kIOHIDSwipeLeft))
                        } else {
                            TouchSimulator.postNavigationSwipeEvent(withDirection: IOHIDSwipeMask(kIOHIDSwipeRight))
                        }
                    }
                    
                    func bfmethod_commandBracket() {
                        DDLogDebug("Actions.swift: NavigationSwipe: Posting bfmethod_commandBracket")
                        let rawKey = isLeft ? kVK_ANSI_LeftBracket : kVK_ANSI_RightBracket
                        let shortcut = MFEmulateNSMenuItemRemapping(CGKeyCode(rawKey), .maskCommand)
                        postKeyboardShortcut(keyCode: shortcut.vkc, modifierFlags: CGSModifierFlags(UInt32(shortcut.modifierMask.rawValue)))
                    }
                    
                    func bfmethod_commandLeftRightArrow() {
                        DDLogDebug("Actions.swift: NavigationSwipe: Posting bfmethod_commandLeftRightArrow")
                        let rawKey = isLeft ? kVK_LeftArrow : kVK_RightArrow
                        postKeyboardShortcut(keyCode: CGKeyCode(rawKey), modifierFlags: CGSModifierFlags(UInt32(CGEventFlags.maskCommand.rawValue)))
                    }
                    
                    func bfmethod_optionCommandBracket() {
                        DDLogDebug("Actions.swift: NavigationSwipe: Posting bfmethod_optionCommandBracket")
                        let rawKey = isLeft ? kVK_ANSI_LeftBracket : kVK_ANSI_RightBracket
                        let mask: CGEventFlags = [.maskAlternate, .maskCommand]
                        let shortcut = MFEmulateNSMenuItemRemapping(CGKeyCode(rawKey), mask)
                        postKeyboardShortcut(keyCode: shortcut.vkc, modifierFlags: CGSModifierFlags(UInt32(shortcut.modifierMask.rawValue)))
                    }
                }
                
            } else if actionType == kMFActionDictTypeSmartZoom {
                TouchSimulator.postSmartZoomEvent()
            } else if actionType == kMFActionDictTypeKeyboardShortcut {
                if let keycodeNum = actionDict[kMFActionDictKeyKeyboardShortcutVariantKeycode] as? NSNumber,
                   let flagsNum = actionDict[kMFActionDictKeyKeyboardShortcutVariantModifierFlags] as? NSNumber {
                    postKeyboardShortcut(keyCode: CGKeyCode(keycodeNum.intValue), modifierFlags: CGSModifierFlags(UInt32(flagsNum.uint64Value)))
                }
            } else if actionType == kMFActionDictTypeSystemDefinedEvent {
                if let typeNum = actionDict[kMFActionDictKeySystemDefinedEventVariantType] as? NSNumber,
                   let flagsNum = actionDict[kMFActionDictKeySystemDefinedEventVariantModifierFlags] as? NSNumber {
                    let type = MFSystemDefinedEventType(rawValue: typeNum.uint32Value)
                    postSystemDefinedEvent(type: type, modifierFlags: NSEvent.ModifierFlags(rawValue: UInt(flagsNum.uint64Value)))
                }
            } else if actionType == kMFActionDictTypeMouseButtonClicks {
                if let buttonNum = actionDict[kMFActionDictKeyMouseButtonClicksVariantButtonNumber] as? NSNumber,
                   let clicksNum = actionDict[kMFActionDictKeyMouseButtonClicksVariantNumberOfClicks] as? NSNumber {
                    ModificationUtility.postMouseButtonClicks(MFMouseButtonNumber(UInt32(buttonNum.intValue)), nOfClicks: Int64(clicksNum.intValue))
                }
            } else if actionType == kMFActionDictTypeAddModeFeedback {
                if let payload = actionDict.mutableCopy() as? NSMutableDictionary {
                    payload.removeObject(forKey: kMFActionDictKeyType)
                    Remap.sendAddModeFeedback(payload)
                }
            } else if actionType == kMFActionDictTypeToggleSmartShift {
                DDLogInfo("[SMARTSHIFT] Actions.swift - ToggleSmartShift action triggered")
                DispatchQueue.main.async {
                    if let activeDev = HelperState.shared.activeDevice {
                        DDLogInfo("[SMARTSHIFT] Actions.swift - ToggleSmartShift: activeDevice is \(activeDev.name) (iohidDevice: \(String(describing: activeDev.iohidDevice)))")
                        if let iohidDevice = activeDev.iohidDevice {
                            let success = LogitechActivator.shared.toggleSmartShiftForDevice(iohidDevice)
                            DDLogInfo("[SMARTSHIFT] Actions.swift - ToggleSmartShift executed. Success: \(success)")
                        }
                    } else {
                        DDLogWarn("[SMARTSHIFT] Actions.swift - ToggleSmartShift failed because no active device was found")
                    }
                }
            }
        }
    }
    
    private static func postSystemDefinedEvent(type: MFSystemDefinedEventType, modifierFlags: NSEvent.ModifierFlags) {
        let tapLoc = CGEventTapLocation.cgSessionEventTap
        let loc = NSEvent.mouseLocation
        
        var data = 0
        data |= Int(kMFSystemDefinedEventBase)
        data |= Int(type.rawValue << 16)
        
        let downData = data
        let upData = data | Int(kMFSystemDefinedEventPressedMask)
        
        let ts = ModificationUtility.nsTimeStamp()
        if let e = NSEvent.otherEvent(with: .systemDefined, location: loc, modifierFlags: modifierFlags, timestamp: ts, windowNumber: -1, context: nil, subtype: 8, data1: downData, data2: -1) {
            if let cgEvent = e.cgEvent {
                cgEvent.post(tap: tapLoc)
            }
        }
        
        let ts2 = ModificationUtility.nsTimeStamp()
        if let e2 = NSEvent.otherEvent(with: .systemDefined, location: loc, modifierFlags: modifierFlags, timestamp: ts2, windowNumber: -1, context: nil, subtype: 8, data1: upData, data2: -1) {
            if let cgEvent2 = e2.cgEvent {
                cgEvent2.post(tap: tapLoc)
            }
        }
    }
    
    private static func postKeyboardShortcut(keyCode: CGKeyCode, modifierFlags: CGSModifierFlags) {
        DDLogDebug("postKeyboardShortcut: Posting shortcut with keyCode: \(keyCode), modifierFlags: \(modifierFlags.rawValue)")
        let tapLoc = CGEventTapLocation.cgSessionEventTap
        
        guard let keyDown = CGEvent(keyboardEventSource: nil, virtualKey: keyCode, keyDown: true),
              let keyUp = CGEvent(keyboardEventSource: nil, virtualKey: keyCode, keyDown: false),
              let modEvent = CGEvent(source: nil) else {
            return
        }
        
        keyDown.flags = CGEventFlags(rawValue: UInt64(modifierFlags.rawValue))
        keyUp.flags = CGEventFlags(rawValue: UInt64(modifierFlags.rawValue))
        modEvent.flags = []
        
        let currentLayout = MFKeyboardTypeCurrent()
        keyDown.setIntegerValueField(.keyboardEventKeyboardType, value: Int64(currentLayout.rawValue))
        keyUp.setIntegerValueField(.keyboardEventKeyboardType, value: Int64(currentLayout.rawValue))
        
        keyDown.post(tap: tapLoc)
        keyUp.post(tap: tapLoc)
        modEvent.post(tap: tapLoc)
    }
}

#endif
